File: apps/driver-android/src/main/kotlin/lk/mageride/driver/safety/SosScreen.kt
package lk.mageride.driver.safety

import androidx.activity.compose.BackHandler
import androidx.annotation.StringRes
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextAlign
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import lk.mageride.driver.R
import lk.mageride.driver.ui.Symbols
import lk.mageride.driver.ui.component.StatusPill
import lk.mageride.driver.ui.component.StatusTone
import lk.mageride.driver.ui.theme.ControlTokens
import lk.mageride.driver.ui.theme.MageRideTheme
import lk.mageride.driver.ui.theme.SosColors
import lk.mageride.shared.data.models.Ulid
import lk.mageride.shared.data.models.iam.EmergencyContact
import lk.mageride.shared.data.models.safety.SosSmsStatus
import org.koin.androidx.compose.koinViewModel
import org.koin.core.parameter.parametersOf

/**
 * **SCR-DA-032 · the driver SOS** (US-12.8, AL-13, D-33).
 *
 * The wireframe on `#2A0A0A`: *"Driver SOS"*, a 128 dp `error` disc inside a translucent halo of
 * itself, the line *"Sending GPS + active trip to emergency contact via SMS…"*, and the contact
 * card carrying *"Amma · +94 77 000 1111"* with a `Sent` pill.
 *
 * **Back cancels while the alarm is armed, and is swallowed while it is in flight.** Popping the
 * screen clears the view model, which cancels the countdown — so a driver who opened this by mistake
 * gets the same outcome from Back as from **Cancel**, which is the safe direction. Once the request
 * has left, there is nothing to go back *to*: `POST /v1/sos` is not revocable, and a gesture that
 * looked like it un-sent an alarm would be the worst affordance on this surface.
 *
 * @param onFinished The driver cancelled before the alarm went, or closed the dispatched state.
 */
@Composable
internal fun SosScreen(rideId: Ulid, onFinished: () -> Unit, modifier: Modifier = Modifier) {
    val viewModel: SosViewModel = koinViewModel { parametersOf(rideId) }
    val state by viewModel.state.collectAsStateWithLifecycle()

    BackHandler(enabled = state.stage == SosStage.SENDING) { /* an alarm in flight is not dismissible */ }

    Surface(modifier = modifier.fillMaxSize(), color = SosColors.background) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(MageRideTheme.spacing.md),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.md, Alignment.CenterVertically),
        ) {
            Text(
                text = stringResource(R.string.sos_title),
                style = MaterialTheme.typography.titleLarge,
                color = SosColors.onSos,
            )

            SosDisc(state = state, onRaise = viewModel::raise)

            Text(
                text = stringResource(statusLine(state)),
                style = MaterialTheme.typography.labelLarge,
                color = if (state.stage == SosStage.FAILED) MaterialTheme.colorScheme.error else SosColors.hint,
                textAlign = TextAlign.Center,
            )

            state.error?.let { message ->
                Text(
                    text = stringResource(message),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.error,
                    textAlign = TextAlign.Center,
                )
            }

            ContactCard(state = state)

            Spacer(modifier = Modifier.size(MageRideTheme.spacing.xs))

            SosFooter(state = state, viewModel = viewModel, onFinished = onFinished)
        }
    }
}

/**
 * The wireframe's disc, and the only control that raises the alarm.
 *
 * While armed it carries the countdown rather than the word: the number is what tells a driver who
 * pressed it by accident that they still have a moment, and D-33's budget is why that moment is
 * three seconds and not ten (see [SosViewModel.COUNTDOWN_SECONDS]).
 */
@Composable
private fun SosDisc(state: SosState, onRaise: () -> Unit, modifier: Modifier = Modifier) {
    Box(
        modifier = modifier
            .size(ControlTokens.SosButton + ControlTokens.SosHalo + ControlTokens.SosHalo)
            .background(SosColors.halo, CircleShape),
        contentAlignment = Alignment.Center,
    ) {
        Surface(
            modifier = Modifier.size(ControlTokens.SosButton),
            shape = CircleShape,
            color = MaterialTheme.colorScheme.error,
            enabled = state.stage == SosStage.ARMED,
            onClick = onRaise,
        ) {
            Box(contentAlignment = Alignment.Center) {
                Text(
                    text = when (state.stage) {
                        SosStage.ARMED -> if (state.awaitingPosition) {
                            SosLabels.SOS
                        } else {
                            state.secondsLeft.toString()
                        }

                        else -> SosLabels.SOS
                    },
                    style = MaterialTheme.typography.headlineMedium,
                    color = MageRideTheme.status.onStatus,
                )
            }
        }
    }
}

/**
 * *"Amma · +94 77 000 1111"* with its `Sent` pill (AL-13).
 *
 * A driver with **no** contact on file is told so here rather than being refused the alarm: the
 * event is still recorded and still reaches the admin live feed, and safety-svc answers
 * `SosSmsStatus.NO_CONTACT` for the leg that has nowhere to go. The fix is SCR-DA-029, which is
 * where the contact is set.
 */
@Composable
private fun ContactCard(state: SosState, modifier: Modifier = Modifier) {
    val contact = state.contact

    Surface(
        modifier = modifier.fillMaxWidth(),
        shape = RoundedCornerShape(MageRideTheme.radius.md),
        color = SosColors.surface,
        border = BorderStroke(ControlTokens.Border, SosColors.outline),
    ) {
        Row(
            modifier = Modifier.padding(MageRideTheme.spacing.sm),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs),
        ) {
            Text(
                text = contact?.let { contactLabel(it) } ?: stringResource(R.string.sos_no_contact),
                modifier = Modifier.weight(1f),
                style = MaterialTheme.typography.bodyMedium,
                color = SosColors.onSos,
            )
            state.smsStatus?.let { StatusPill(label = stringResource(smsLabel(it)), tone = smsTone(it)) }
        }
    }
}

/** Cancel while armed; Close once the alarm has gone; Try again when it never left the handset. */
@Composable
private fun SosFooter(
    state: SosState,
    viewModel: SosViewModel,
    onFinished: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Box(modifier = modifier) {
        when (state.stage) {
            SosStage.ARMED -> TextButton(
                onClick = {
                    viewModel.cancelCountdown()
                    onFinished()
                },
            ) {
                Text(text = stringResource(R.string.action_cancel), color = SosColors.onSos)
            }

            SosStage.SENDING -> Text(
                text = stringResource(R.string.sos_sending),
                style = MaterialTheme.typography.labelLarge,
                color = SosColors.hint,
            )

            SosStage.DISPATCHED -> TextButton(onClick = onFinished) {
                Text(text = stringResource(R.string.action_close), color = SosColors.onSos)
            }

            SosStage.FAILED -> Row(horizontalArrangement = Arrangement.spacedBy(MageRideTheme.spacing.xs)) {
                TextButton(onClick = viewModel::retry) {
                    Text(text = stringResource(R.string.action_retry), color = SosColors.onSos)
                }
                TextButton(onClick = onFinished) {
                    Text(text = stringResource(R.string.action_close), color = SosColors.hint)
                }
            }
        }
    }
}

/**
 * The word on the disc, and why it is not copy.
 *
 * `SOS` is an international distress signal, not a sentence: it is the same three letters in
 * Sinhala, Tamil and English, and three identical values in the three `strings.xml` files is exactly
 * what `StringResourceTest` (correctly) reads as a key nobody translated. Same rule as `Rs`, `L3`
 * and `+94` (C068). The *title* above it is ordinary copy and is translated.
 */
internal object SosLabels {
    const val SOS: String = "SOS"
}

/** *"Amma · +94 77 000 1111"*. The number is a proper noun and the dot is [Symbols.DOT]. */
private fun contactLabel(contact: EmergencyContact): String = "${contact.name} ${Symbols.DOT} ${contact.phone}"

/** The line under the disc. Four states, four sentences. */
@StringRes
private fun statusLine(state: SosState): Int = when (state.stage) {
    SosStage.ARMED -> if (state.awaitingPosition) R.string.sos_waiting_position else R.string.sos_armed
    SosStage.SENDING -> R.string.sos_sending_body
    SosStage.DISPATCHED -> R.string.sos_dispatched
    SosStage.FAILED -> R.string.sos_failed
}

/**
 * What the pill says about D-33's parallel gateways.
 *
 * `Failed` is **not** an error state on this screen and does not colour like one: the alert is
 * recorded and is on the admin live feed either way, and telling somebody in trouble that nothing
 * happened would be worse than telling them the SMS leg did not manage it.
 */
@StringRes
private fun smsLabel(status: SosSmsStatus): Int = when (status) {
    SosSmsStatus.DISPATCHED -> R.string.sos_sms_sent
    SosSmsStatus.FAILED -> R.string.sos_sms_failed
    SosSmsStatus.NO_CONTACT -> R.string.sos_sms_no_contact
}

private fun smsTone(status: SosSmsStatus): StatusTone = when (status) {
    SosSmsStatus.DISPATCHED -> StatusTone.DONE
    SosSmsStatus.FAILED, SosSmsStatus.NO_CONTACT -> StatusTone.PENDING
}

File: apps/driver-ios/DriverApp/Safety/SosScreen.swift
import MageRideShared
import SwiftUI

/// **SCR-DI-032 · the driver SOS** (US-12.8, AL-13, D-33).
///
/// The wireframe on `#2A0A0A`: *"Driver SOS"*, a 128pt `error` disc inside a translucent halo of
/// itself, the line *"Sending GPS + active trip via SMS…"*, and the contact card carrying
/// *"Amma · +94 77 000 1111"* with a `Sent` pill.
///
/// **A takeover, not a destination.** ``DriverRoute/sos(rideId:)`` is
/// ``DriverRoute/isFullScreenTakeover``, so this is presented over the whole app, tab bar included —
/// a driver on an alarm screen must not be one tap from their wallet.
///
/// **There is nothing to swallow, which is a real Section C difference.** The Android twin disables
/// its `BackHandler` while the request is in flight, because a hardware Back would otherwise pop the
/// screen and clear the view model mid-`POST`. `fullScreenCover` has no interactive dismissal on
/// iOS: the only way out is the button this screen draws, so *"an alarm in flight is not
/// dismissible"* is the presentation's own property here rather than a handler. The same reasoning
/// ``UpdateGate`` records for its `.alert`.
///
/// - Parameter onFinished: The driver cancelled before the alarm went, or closed the dispatched
///   state. Closes the takeover.
@MainActor
struct SosScreen: View {

    @StateObject private var model: SosModel

    private let onFinished: () -> Void

    init(model: @autoclosure @escaping () -> SosModel, onFinished: @escaping () -> Void) {
        _model = StateObject(wrappedValue: model())
        self.onFinished = onFinished
    }

    var body: some View {
        VStack(spacing: MageRideSpacing.md) {
            Spacer(minLength: 0)

            Text(key: "sos_title")
                .mageFont(.title)
                .foregroundStyle(MageRideSosColor.onSos)

            disc

            Text(key: statusKey)
                .mageFont(.label)
                .foregroundStyle(model.state.stage == .failed ? MageRideColor.error : MageRideSosColor.hint)
                .multilineTextAlignment(.center)

            if let errorKey = model.state.errorKey {
                Text(key: errorKey)
                    .mageFont(.bodySmall)
                    .foregroundStyle(MageRideColor.error)
                    .multilineTextAlignment(.center)
            }

            contactCard

            Spacer(minLength: 0)

            footer
        }
        .padding(MageRideSpacing.md)
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(MageRideSosColor.background)
        .task { model.start() }
        .onDisappear(perform: model.stop)
    }

    /// The wireframe's disc, and the only control that raises the alarm.
    ///
    /// While armed it carries the **countdown** rather than the word: the number is what tells a
    /// driver who pressed it by accident that they still have a moment, and D-33's budget is why
    /// that moment is three seconds and not ten (see ``SosModel/countdownSeconds``).
    private var disc: some View {
        ZStack {
            Circle()
                .fill(MageRideSosColor.halo)
                .frame(
                    width: MageRideControl.sosButton + MageRideControl.sosHalo * 2,
                    height: MageRideControl.sosButton + MageRideControl.sosHalo * 2
                )

            Button(action: model.raise) {
                Text(discLabel)
                    .mageFont(.headline)
                    .foregroundStyle(MageRideColor.onStatus)
                    .frame(width: MageRideControl.sosButton, height: MageRideControl.sosButton)
                    .background(MageRideColor.error, in: Circle())
                    .contentShape(Circle())
            }
            .buttonStyle(.plain)
            .disabled(model.state.stage != .armed)
        }
        // One announcement, and it is the *action* rather than the number: a reader hearing "3"
        // learns nothing, and a driver using VoiceOver in an emergency needs the button named.
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(Text(key: "sos_title"))
        .accessibilityValue(Text(key: statusKey))
        .accessibilityAddTraits(.isButton)
    }

    /// *"Amma · +94 77 000 1111"* with its `Sent` pill (AL-13).
    ///
    /// A driver with **no** contact on file is told so here rather than being refused the alarm: the
    /// event is still recorded and still reaches the admin live feed, and safety-svc answers
    /// `SosSmsStatus.noContact` for the leg that has nowhere to go. The fix is SCR-DI-029, which is
    /// where the contact is set.
    private var contactCard: some View {
        HStack(spacing: MageRideSpacing.xs) {
            Text(contactLabel)
                .mageFont(.bodySmall)
                .foregroundStyle(MageRideSosColor.onSos)
                .frame(maxWidth: .infinity, alignment: .leading)

            if let status = model.state.smsStatus {
                StatusPill(label: Self.smsLabelKey(status).localised, tone: Self.smsTone(status))
            }
        }
        .padding(MageRideSpacing.sm)
        .frame(maxWidth: .infinity)
        .background(
            MageRideSosColor.surface,
            in: RoundedRectangle(cornerRadius: MageRideRadius.md, style: .continuous)
        )
        .overlay {
            RoundedRectangle(cornerRadius: MageRideRadius.md, style: .continuous)
                .strokeBorder(MageRideSosColor.outline, lineWidth: 1)
        }
        .accessibilityElement(children: .combine)
    }

    /// Cancel while armed; Close once the alarm has gone; Try again when it never left the handset.
    @ViewBuilder
    private var footer: some View {
        switch model.state.stage {
        case .armed:
            Button {
                model.cancelCountdown()
                onFinished()
            } label: {
                Text(key: "action_cancel")
                    .mageFont(.subtitle)
                    .foregroundStyle(MageRideSosColor.onSos)
                    .frame(maxWidth: .infinity, minHeight: MageRideControl.minimumTapTarget)
            }
            .buttonStyle(.plain)

        case .sending:
            Text(key: "sos_sending")
                .mageFont(.label)
                .foregroundStyle(MageRideSosColor.hint)
                .frame(minHeight: MageRideControl.minimumTapTarget)

        case .dispatched:
            Button(action: onFinished) {
                Text(key: "action_close")
                    .mageFont(.subtitle)
                    .foregroundStyle(MageRideSosColor.onSos)
                    .frame(maxWidth: .infinity, minHeight: MageRideControl.minimumTapTarget)
            }
            .buttonStyle(.plain)

        case .failed:
            HStack(spacing: MageRideSpacing.md) {
                Button(action: model.retry) {
                    Text(key: "action_retry")
                        .mageFont(.subtitle)
                        .foregroundStyle(MageRideSosColor.onSos)
                        .frame(maxWidth: .infinity, minHeight: MageRideControl.minimumTapTarget)
                }
                .buttonStyle(.plain)

                Button(action: onFinished) {
                    Text(key: "action_close")
                        .mageFont(.subtitle)
                        .foregroundStyle(MageRideSosColor.hint)
                        .frame(maxWidth: .infinity, minHeight: MageRideControl.minimumTapTarget)
                }
                .buttonStyle(.plain)
            }
        }
    }

    /// What the disc reads — the countdown while there is one to run, the distress signal otherwise.
    private var discLabel: String {
        guard model.state.stage == .armed, !model.state.isAwaitingPosition else { return SosLabels.sos }
        return String(model.state.secondsLeft)
    }

    /// *"Amma · +94 77 000 1111"*. The number is a proper noun and the dot is ``MageRideSymbols``'.
    private var contactLabel: String {
        guard let contact = model.state.contact else { return "sos_no_contact".localised }
        return contact.name + MageRideSymbols.separator + contact.phone
    }

    /// The line under the disc. Four states, four sentences.
    private var statusKey: String {
        switch model.state.stage {
        case .armed: return model.state.isAwaitingPosition ? "sos_waiting_position" : "sos_armed"
        case .sending: return "sos_sending_body"
        case .dispatched: return "sos_dispatched"
        case .failed: return "sos_failed"
        }
    }

    /// What the pill says about D-33's parallel gateways.
    ///
    /// `Failed` is **not** an error state on this screen and does not colour like one: the alert is
    /// recorded and is on the admin live feed either way, and telling somebody in trouble that
    /// nothing happened would be worse than telling them the SMS leg did not manage it.
    static func smsLabelKey(_ status: SosSmsStatus) -> String {
        switch status {
        case SosSmsStatus.dispatched: return "sos_sms_sent"
        case SosSmsStatus.failed: return "sos_sms_failed"
        default: return "sos_sms_no_contact"
        }
    }

    static func smsTone(_ status: SosSmsStatus) -> StatusTone {
        status == SosSmsStatus.dispatched ? .done : .pending
    }
}

/// The word on the disc, and why it is not copy.
///
/// `SOS` is an international distress signal, not a sentence: it is the same three letters in
/// Sinhala, Tamil and English, and three identical values in the three `Localizable.strings` files is
/// exactly what `LocalizationTests.testNoTranslationWasLeftAsItsEnglishPlaceholder` reads as a key
/// nobody translated. Same rule as `Rs`, `L3` and `+94` (C086), and the same constant as
/// `apps/driver-android/.../safety/SosScreen.kt`'s `SosLabels`. The **title** above it is ordinary
/// copy and is translated.
enum SosLabels {
    static let sos = "SOS"
}
